src/syntax/FStar.Syntax.DsEnv.fst

```diff
@@ -1041,7 +1045,128 @@ let push_sigelt' fail_on_dup env s =
 let push_sigelt       env se = push_sigelt' true  env se
 let push_sigelt_force env se = push_sigelt' false env se
 
-let push_namespace env ns =
+let find_data_constructors_for_typ env (lid:lident) =
+  let k_global_def lid = function
+      | ({ sigel = Sig_inductive_typ {ds} }, _) -> Some ds
+      | _ -> None in
+  resolve_in_open_namespaces' env lid (fun _ -> None) (fun _ -> None) k_global_def
+
+let find_binders_for_datacons env (lid:lident) =
+  let k_global_def lid = function
+      | ({ sigel = Sig_datacon {t} }, _) ->
+          arrow_formals_comp_ln t
+        |> fst
+        |> List.map (fun x -> x.binder_bv.ppname)
+        |> Some
+      | _ -> None in
+  resolve_in_open_namespaces' env lid (fun _ -> None) (fun _ -> None) k_global_def
+
+// let x (xx: (lident & list term)) = show xx
+
+
+(** Elaborates a `restriction`: this function adds implicit names
+(projectors, discriminators, record fields) that F* generates
+automatically. It also checks that all the idents the user added
+actually exists in the given namespace. *)
+let elab_restriction f env ns restriction =
+  let open FStar.Class.Deq in
+  match restriction with
+  | Unrestricted -> f env ns restriction
+  | AllowList l  ->
+    let mk_lid (id: ident): lident = set_lid_range (lid_of_ids (ids_of_lid (qual_id ns id))) (range_of_id id) in
+    let name_exists id =
+      let lid = mk_lid id in
+      match try_lookup_lid env lid with
+      | Some _ -> true
+      | None   -> try_lookup_record_or_dc_by_field_name env lid |> is_some
+    in
+    // For every inductive, we include its constructors
+    let l = List.map (fun (id, renamed) ->
+      let with_id_range = dflt id renamed |> range_of_id |> set_id_range in
+        match find_data_constructors_for_typ env (mk_lid id) with
+      | Some idents -> List.map (fun id -> (ident_of_lid id |> with_id_range, None)) idents
+      | None -> []
+    ) l |> List.flatten |> List.append l in
+    // For every constructor, we include possible desugared record
+    // payloads types
+    let l =
+      (* A (precomputed) associated list that maps a constructors to
+         types that comes from a "record-on-a-variant" desugar. E.g. `A`
+         is mapped to `Mka__A__payload` for a `type a = | A {x:int}`. *)
+      let constructor_lid_to_desugared_record_lids: list (ident * ident) =
+        env.modules
+        |> List.concatMap (fun (_, {declarations}) -> declarations)
+        |> List.concatMap (fun sigelt -> match sigelt.sigel with | Sig_bundle {ses} -> ses | _ -> [])
+        |> List.concatMap (fun sigelt -> lids_of_sigelt sigelt |> List.map (fun lid -> (lid, sigelt.sigattrs)))
+        |> List.filter_map
+          (fun (lid, attrs) -> match U.get_attribute Const.desugar_of_variant_record_lid attrs with
+                            | Some [({n = Tm_constant (FStar.Const.Const_string (s, _))}, None)]
```

Comment:
Similarly to my other comment `unembed` could make this a bit neater maybe

Reply:
Sadly here that's causing a cycle :'(